### ISSUE TITLE
cmdline autocompletion

### ISSUE THREAD

**Issue by deathbeam** [nvim-mini/mini.nvim#690]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

### Module(s)

mini.completion

### Description

I recently switched to mini.completion from nvim-cmp and I feel like there must be simpler way to do command line autocomplete that isnt something like wilder. I got fairly close with just autocommands, only issue was wildmenu behaviour that cant rly be changed where it always tries to complete if there is only match (so something like noinsert) is not possible. Could be good thing to look at maybe for mini as it do not sounds like it would require a lot of code.

**Comment by echasnovski** [comment]
Thanks for the suggestion!

I did not look into Command line completion, as the built-in default one seems good enough for me.

Would you mind elaborating on which issues are worth solving here?

**Comment by deathbeam** [comment]
Well its mostly that it would be nice to have the menu open as you type instead of having to press tab, nothing fancy (see https://github.com/gelguy/wilder.nvim or https://github.com/hrsh7th/cmp-cmdline for nvim-cmp), and example:

![image](https://github.com/echasnovski/mini.nvim/assets/5115805/a1f98da6-adfc-4854-9780-048bfc364a48)

**Comment by echasnovski** [comment]
Ah, so basically just make a Command line autocompletion (as the title says :) ). Yeah, sounds interesting. I'll see what can be done when I'll be updating 'mini.completion' in (hopefully) near-ish future.

**Comment by deathbeam** [comment]
Alright so I was pretty bored and found [this](https://github.com/smolck/command-completion.nvim) as example and tried to implement it properly here: https://github.com/deathbeam/dotfiles/blob/master/nvim/.config/nvim/lua/config/cmdline-completion.lua .

![image](https://github.com/echasnovski/mini.nvim/assets/5115805/3db249e9-0cf2-48f1-a423-06283ef6aff1)

Im not sure if you would be willing to accept this as PR (or if you wanted it merged with mini.completion) but I would be happy to PR it if you do.

**Comment by echasnovski** [comment]
Huge thanks for the suggestion and enthusiasm!

At the moment I would like to take time to think about the best approach here. Whether it should be added in 'mini.completion' or separate module (or even at all, really).

**Comment by Yushi5058** [comment]
Sounds like a good feature, I'm waiting for updates !!

**Comment by GNOMES** [comment]
If it doesn't add too many lines to make mini.completion no longer "mini", it would be great to have one extension that could replace three CMP plugins (cmp/cmp-cmdline/cmp-buffer)
.

**Comment by echasnovski** [comment]
Indeed, my main course of action here is to see if it can be included in 'mini.completion'.

**Comment by echasnovski** [comment]
I've looked into this and I *think* it would be a better fit for the future 'mini.cmdline'. But not yet sure.

**Comment by deathbeam** [comment]
After this: <https://github.com/neovim/neovim/pull/32709> making autocompletion for cmdline is very easy as well (see the example that I made for that PR there as well, thats almost everything that you need to do there) instead of what had to be done previously e.g basically recreating the whole wildmenu

**Comment by 231tr0n** [comment]
```lua
    vim.opt.wildmenu = true
    vim.opt.wildoptions = 'pum,fuzzy'
    vim.opt.wildmode = 'noselect:lastused,full'
    vim.opt.wildcharm = vim.fn.char2nr(term)

    vim.keymap.set('c', '<Up>', '<End><C-U><Up>', { silent = true })
    vim.keymap.set('c', '<Down>', '<End><C-U><Down>', { silent = true })

    vim.api.nvim_create_autocmd('CmdlineChanged', {
        pattern = ':',
        callback = function()
            local cmdline = vim.fn.getcmdline()
            local curpos = vim.fn.getcmdpos()
            local last_char = cmdline:sub(-1)

            if
                curpos == #cmdline + 1
                and vim.fn.pumvisible() == 0
                and last_char:match('[%w%/%: ]')
                and not cmdline:match('^%d+$')
            then
                vim.opt.eventignore:append('CmdlineChanged')
                vim.api.nvim_feedkeys(term, 'ti', false)
                vim.schedule(function()
                    vim.opt.eventignore:remove('CmdlineChanged')
                end)
            end
        end,
    })
```

This worked for me. Thank you @deathbeam

**Comment by luisdavim** [comment]
I find the search completion in cmp really useful.

**Comment by rawsp33d** [comment]
@deathbeam and @231tr0n, the code you suggested seemed to be working great, until I discovered that it breaks typing certain characters in commands. The two I have noticed so far are `/` and `<space>`.

For example, if I type `:s/` it gets replaced with `:s/^Z` and I cannot delete the `^Z`. I was having issues with spaces getting replaced by `^Z` too but I can't seem to recreate that.

I got it working by conditionally feeding `<c-z>` based on whether the last character was a slash. It feels hacky and there may well be other characters it chokes on, but maybe if we can understand why we can start fixing it more generally and less hackily. I also used `vim.keycode` instead of `vim.api.nvim_replace_termcodes` which I _think_ is equivalent to the `, true, true, true` arguments but am not sure.

```lua
local term = vim.keycode("<c-z>")
vim.opt.wildoptions = "pum,fuzzy"
vim.opt.wildmode = "noselect:lastused,full"
vim.opt.wildcharm = vim.fn.char2nr(term)

-- trigger completion when typing in command mode
vim.api.nvim_create_autocmd("CmdlineChanged", {
	pattern = ":",
	callback = function()
		local cmdline = vim.fn.getcmdline()
		local curpos = vim.fn.getcmdpos()
		local last_char = cmdline:sub(-1)
		if
				curpos == #cmdline + 1
				and vim.fn.pumvisible() == 0
				and last_char:match("[%w%/%: ]")
				and last_char ~= "/"
				and not cmdline:match("^%d+$")
		then
			vim.opt.eventignore:append("CmdlineChanged")
			vim.api.nvim_feedkeys(term, "ti", false)
			vim.schedule(function()
				vim.opt.eventignore:remove("CmdlineChanged")
			end)
		end
	end,
})
```

**Comment by deathbeam** [comment]
On latest neovim you can just do this and it works ™️ :

```lua
    local group = vim.api.nvim_create_augroup('myplugins-cmdcomplete', { clear = true })

    vim.opt.wildmode = 'noselect:lastused,full'
    vim.keymap.set('c', '<Up>', '<End><C-U><Up>', { silent = true })
    vim.keymap.set('c', '<Down>', '<End><C-U><Down>', { silent = true })

    vim.api.nvim_create_autocmd('CmdlineChanged', {
        group = group,
        desc = 'Auto show command line completion',
        pattern = ':',
        callback = function()
            vim.fn.wildtrigger()
        end,
    })
```

(e.g use the new vim.fn.wildtrigger() ported from vim)

**Comment by echasnovski** [comment]
> On latest neovim you can just do this and it works ™️ :

I did try it for couple of days, but had several issues. Like flickering (pmenu seems to disappear and momentarily appear) and not fully working arrows (but I prepended them only with `<C-u>`, as `:h wildtrigger()` suggests). But maybe it indeed looks like an interesting solution worth tweaking.

**Comment by echasnovski** [comment]
Command line autocompletion is now a part of 'mini.nvim' via ['mini.cmdline'](https://nvim-mini.org/mini.nvim/readmes/mini-cmdline.html). It is basically a [more customizable version of `:h cmdline-autocompletion` on Neovim>=0.12](https://github.com/nvim-mini/mini.nvim/blob/e27fcecf6cbffd8ed391bdf0eb746fdb5b9bfa44/doc/mini-cmdline.txt#L58-L60) which also works on Neovim<0.12.
